client/components/popover/index.jsx

```diff
@@ -81,245 +86,176 @@ class Popover extends Component {
 
 	/**
 	 * Flag to determine if we're currently repositioning the Popover
+	 *
 	 * @type {boolean} True if the Popover is being repositioned.
 	 */
 	isUpdatingPosition = false;
 
-	constructor( props ) {
-		super( props );
-
-		this.setPopoverId( props.id );
+	popoverNodeRef = React.createRef();
+	popoverInnerNodeRef = React.createRef();
 
-		// bound methods
-		this.setDOMBehavior = this.setDOMBehavior.bind( this );
-		this.setPosition = this.setPosition.bind( this );
-		this.onClickout = this.onClickout.bind( this );
-		this.onKeydown = this.onKeydown.bind( this );
-		this.onWindowChange = this.onWindowChange.bind( this );
-
-		this.state = {
-			show: props.isVisible,
-			left: -99999,
-			top: -99999,
-			positionClass: this.getPositionClass( props.position ),
-		};
-	}
+	state = {
+		show: this.props.isVisible,
+		left: -99999,
+		top: -99999,
+		positionClass: this.getPositionClass( this.props.position ),
+	};
 
 	componentDidMount() {
 		if ( this.state.show ) {
-			this.bindEscKeyListener();
-			this.bindDebouncedReposition();
-			bindWindowListeners();
+			this.bindListeners();
+			this.setPositionAndFocus();
 		}
 	}
 
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		// update context (target) reference into a property
-		this.domContext = ReactDom.findDOMNode( nextProps.context );
-
-		if ( ! nextProps.isVisible ) {
-			return null;
-		}
-
-		this.setPosition();
-	}
-
 	componentDidUpdate( prevProps, prevState ) {
-		const { isVisible } = this.props;
-
-		if ( ! prevState.show && this.state.show ) {
-			this.bindEscKeyListener();
-			this.bindDebouncedReposition();
-			bindWindowListeners();
-		}
-
-		if ( isVisible !== prevProps.isVisible ) {
-			if ( isVisible ) {
+		// Show can be delayed with `showDelay`, so `state.show` lags behind `props.isVisible`
+		if ( this.props.isVisible !== prevProps.isVisible ) {
+			if ( this.props.isVisible ) {
 				this.show();
 			} else {
 				this.hide();
 			}
 		}
 
-		if ( ! this.domContainer || ! this.domContext ) {
-			return null;
+		// When showing, bind window listeners, set position and focus.
+		// Corresponds exactly to `componentDidMount`.
+		if ( ! prevState.show && this.state.show ) {
+			this.bindListeners();
+			this.setPositionAndFocus();
```

Comment:
Will this not invalidate layout, because of the transient `computePosition` call? If I recall correctly, `componentDidUpdate` is a good time for DOM manipulations, but not a great time for DOM reads.

Reply:
I think it's the opposite: good time for DOM reads, because the layout is already done and reading from DOM doesn't trigger any flush. If we write to DOM at this time, a second layout will be triggered, but there's no way around it: we need a two-pass layout here. First we layout the popover, and in the second pass we measure its layout and decide about the final layout (positioning) there.

In your #37489, you're postponing DOM reads to `componentDidMount`, where previously they were reading at the time when ref callbacks are running. Which, according to your profiling results, is too early and causes layout trashing.

Reply:
Right, the `ref` callback time is another thing altogether, which I'm not sure exactly how it fit. But regardless of that, you're right that in #37489 using the `componentDidMount` hook seemed to work well for DOM reads, which means I was confused, yup.

The two-pass layout is a shame, but a common issue in these types of components.

In any case, I'll give this a performance look-over with the full review. Teaches me to start writing comments before I measure things 🙂